### PR TITLE
feat(tracer): support methods with the same name (ABCs) by including fully qualified name in v2

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #1454 

### Changes

> Please provide a summary of what's being changed

This change modifies the name of the range generated by the capture_method tracker. In some cases it is very difficult to identify which functions are added in XRay just using the function name, the reason for this is because a python function/file/system can use abstract methods/classes or even functions with the same name in different classes.

image

### User experience

> Please share what the user experience looks like before and after this change

Before to this change, the span name only used the function name:
image

After this change, the span name includes the module and Python qualified name ([PEP 3155](https://peps.python.org/pep-3155/))

image

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/tracing/tracer.py [hunk 1]
```diff
@@ -508,7 +508,7 @@ async def async_tasks():
                 functools.partial(self.capture_method, capture_response=capture_response, capture_error=capture_error),
             )
 
-        method_name = f"{method.__name__}"
+        method_name = f"{method.__module__}.{method.__qualname__}"
```
Comment (heitorlessa) [inline aws_lambda_powertools/tracing/tracer.py:513]: Add a comment above this line so we remember the expected format.

File: tests/unit/test_tracing.py [hunk 2]
```diff
@@ -125,9 +125,11 @@ def greeting(name, message):
     # and add its response as trace metadata
     # and use service name as a metadata namespace
     assert in_subsegment_mock.in_subsegment.call_count == 1
-    assert in_subsegment_mock.in_subsegment.call_args == mocker.call(name="## greeting")
+    assert in_subsegment_mock.in_subsegment.call_args == mocker.call(
+        name="## unit.test_tracing.test_tracer_method.<locals>.greeting"
+    )
     assert in_subsegment_mock.put_metadata.call_args == mocker.call(
-        key="greeting response", value=dummy_response, namespace="booking"
+        key="unit.test_tracing.test_tracer_method.<locals>.greeting response", value=dummy_response, namespace="booking"
```
Comment (heitorlessa) [inline tests/unit/test_tracing.py]: make it as a constant and use f-string to make it easier to read/maintain later.

File: tests/e2e/tracer/handlers/same_function_name.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    @tracer.capture_method
+    def get_all(self):
+        return [{"id": f"{uuid4()}", "completed": False} for _ in range(5)]
+
+
+class Todos:
+    @tracer.capture_method
+    def get_all(self):
+        return [{"id": f"{uuid4()}", "completed": False} for _ in range(5)]
+
```
Comment (heitorlessa) [inline tests/e2e/tracer/handlers/same_function_name.py]: Create an ABC so it's closer to what customers reported to us when they hit this issue.

File: tests/e2e/tracer/test_tracer.py [hunk 4]
```diff
@@ -46,14 +56,39 @@ def test_lambda_handler_trace_is_visible(basic_handler_fn_arn: str, basic_handle
...
+
+    method_name_2 = "same_function_name.Comments.get_all"
+    method_subsegment_2 = f"## {method_name_2}"
+    method_metadata_key_2 = f"{method_name_2} response"
+
+    trace_query = data_builder.build_trace_default_query(function_name=same_function_name_fn)
+
+    # WHEN
+    _, execution_time = data_fetcher.get_lambda_response(lambda_arn=same_function_name_arn)
+    data_fetcher.get_lambda_response(lambda_arn=same_function_name_arn)
```
Comment (heitorlessa) [inline tests/e2e/tracer/test_tracer.py]: no need as we're already asserting on Cold Start in another test.

File: tests/e2e/tracer/test_tracer.py [hunk 5]
```diff
@@ -46,14 +56,39 @@ def test_lambda_handler_trace_is_visible(basic_handler_fn_arn: str, basic_handle
...
 
+def test_lambda_handler_trace_multiple_functions_same_name(same_function_name_arn: str, same_function_name_fn: str):
+    # GIVEN
+    method_name_1 = "same_function_name.Todos.get_all"
+    method_subsegment_1 = f"## {method_name_1}"
+    method_metadata_key_1 = f"{method_name_1} response"
+
+    method_name_2 = "same_function_name.Comments.get_all"
+    method_subsegment_2 = f"## {method_name_2}"
+    method_metadata_key_2 = f"{method_name_2} response"
```
Comment (heitorlessa) [inline tests/e2e/tracer/test_tracer.py]: make the names explicit like `todo` and `comments`

**Review by heitorlessa — CHANGES_REQUESTED** [review]
looks great!! Made a few comments to enhance maintenance.

File: aws_lambda_powertools/tracing/tracer.py (hunk 1 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/tracing/tracer.py:513]: Hi @heitorlessa! Hmm probably not a good idea. Flake mark it as possible commented code and adding an exception can make the code confusing.

```
linting::flake8..........................................................Failed
- hook id: flake8
- exit code: 1

aws_lambda_powertools/tracing/tracer.py:512:1: E800 Found commented out code
```

Instead, I updated the docstring:
image

Pls let me know what do you think about.

File: tests/e2e/tracer/handlers/same_function_name.py (hunk 3 above)
Comment (leandrodamascena) [reply tests/e2e/tracer/handlers/same_function_name.py]: Looks great now :rocket: 
image

File: tests/e2e/tracer/test_tracer.py (hunk 4 above)
Comment (leandrodamascena) [reply tests/e2e/tracer/test_tracer.py]: Removed

File: tests/e2e/tracer/test_tracer.py (hunk 5 above)
Comment (leandrodamascena) [reply tests/e2e/tracer/test_tracer.py]: renamed

File: tests/unit/test_tracing.py (hunk 2 above)
Comment (leandrodamascena) [reply tests/unit/test_tracing.py]: Now it's a constant

image

File: aws_lambda_powertools/tracing/tracer.py (hunk 1 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/tracing/tracer.py:513]: Hello @heitorlessa! I added the comment above the suggested line.

**Comment by heitorlessa** [comment]
Looking...

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**